x-pack/legacy/plugins/index_management/static/ui/components/mappings_editor.tsx

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { useState, useEffect } from 'react';
+import { EuiCodeEditor } from '@elastic/eui';
+
+interface Props {
+  setGetDataHandler: (handler: () => { isValid: boolean; data: Mappings }) => void;
```

Comment:
This is on purpose _not_ an onChange handler. We don't want to parse (`JSON.parse`) on each key stroke (which unfortunately we do in other places). The same optimization is needed for the future editor (we don't want to build the object and validate all the fields on each key stroke).
The parent component decides when he needs the data (probably when sending the form) and ask for it once.

Reply:
> We don't want to parse (JSON.parse) on each key stroke 

Can we delegate that to the consumer? I think it's OK to let the consumer own that aspect of validation and incur a breaking change once we switch the editor from managing a JSON string to managing an object. This would also simplify this component because then it wouldn't need the JSON error callout.

> we don't want to build the object and validate all the fields on each key stroke

Building on top of my suggestion, have you considered calling the `onChange` callback with an object with this shape?

```js
{
  data, // the data object or JSON string managed by the mapping editor
  validate, // a function the consumer can call to validate the data
}
```

I think your requirement makes sense, and it seems like this would satisfy it?

Reply:
I am not sure I follow your reasoning. Can we Zoom so you can explain to me better your concerns on this placeholder component that will be replaced before the release?

Reply:
We Zoomed about this and identified a couple goals, which are sometimes at odds with one another:

1. Leveraging well-known concepts like `onChange` makes it easier for people to understand how to consume components.
2. Performance considerations can conflict with this goal. In this case, the mappings editor will eventually use a special optimized format for storing the form state which wouldn't make sense to expose to the consumer.

We agreed to keep our options open and see if the performance optimization is necessary as we move forward. I'm OK with merging this as-is. Thanks for the chat @sebelga!